### PR TITLE
Adding HEAD for heartbeat api calls. Issue #165

### PR DESCRIPTION
**Pull request by bsieber-mozilla**
Allow HEAD as well as GET requests to :

    /__heartbeat__
    /__lb_heartbeat__

### CODE REVIEW

File: ctms/sample_data.py [hunk 1]
```diff
@@ -177,7 +177,8 @@ class ContactVendor:
     }
 
     def __getitem__(self, key: str) -> ContactSchema:
-        return self.contacts[key][0].copy(deep=True)
+        contact: ContactSchema = self.contacts[key][0].copy(deep=True)
+        return contact
```
Comment (bsieber-mozilla) [inline ctms/sample_data.py:181]: > ctms/sample_data.py:180: error: Returning Any from function declared to return "ContactSchema"
> Found 1 error in 1 file (checked 2 source files)
Fix for this mypy pre-commit issue which came up?

File: tests/unit/test_metrics.py [hunk 2]
```diff
@@ -15,10 +15,10 @@
 # Higher numbers = more ways to slice data, more storage, more processing time for summaries
 
 # Cardinality of ctms_requests_total counter
-METHOD_PATH_CODE_COMBINATIONS = 45
+METHOD_PATH_CODE_COMBINATIONS = 47
 
 # Cardinality of ctms_requests_duration_seconds histogram
-METHOD_PATH_CODEFAM_COMBOS = 31
+METHOD_PATH_CODEFAM_COMBOS = 33
 DURATION_BUCKETS = 8
```
Comment (bsieber-mozilla) [inline tests/unit/test_metrics.py:22]: With 2 new HEAD methods, is this correctly adjusted...?

File: tests/unit/test_api.py [hunk 3]
```diff
@@ -410,30 +414,47 @@ def test_unauthorized_api_call_fails(
...
+        "GET": ({"method": "GET", "url": path, "params": params}, {200}),
+        "HEAD": ({"method": "HEAD", "url": path}, {200}),
+        "PUT": (put_patch_post("PUT"), {200, 201}),  # Either creates or updates
+        "PATCH": (put_patch_post("PATCH"), {200}),
+        "POST": (put_patch_post("POST"), {201}),
+    }
+    if method in _map.keys():
+        _request, _results = _map.get(method)
+        resp = client.request(**_request)
+        assert resp.status_code in _results
```
[the remark this answers is not in the document]
Comment (jwhitlock) [reply tests/unit/test_api.py:458]: Yes, we can split up this file. I hit this warning when adding ``PATCH``, so I put all those tests in ``test_api_patch.py``. A similar method could be used to break out the ``POST`` and ``PUT`` tests into their own files.

I've been testing monitoring endpoints in ``test_app.py``, I think these new checks should go there as well, which will keep us under 1000 for another day.

File: tests/unit/test_metrics.py (hunk 2 above)
Comment (jwhitlock) [reply tests/unit/test_metrics.py:22]: That appears correct to me, adds two new method/path combos

File: ctms/sample_data.py (hunk 1 above)
Comment (jwhitlock) [reply ctms/sample_data.py:181]: Maybe! I don't see this error in my dev environment, however, so I can't tell for sure.

The sample data needs to be moved out of ``ctms`` and into ``tests/unit``, but that's for a future PR.

**Review by jwhitlock — APPROVED** [review]
Thanks @bsieber-mozilla! I think the code change looks good, but the tests should be in ``tests_app.py`` or a new file ``tests_monitor.py``, not in ``tests_api.py``. 

There's some stalled work to add this to FastAPI in https://github.com/tiangolo/fastapi/issues/1773. I don't think we should wait for it.

File: tests/unit/test_monitor.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+def test_authorized_api_call_succeeds(client, example_contact, method, path, params):
+    """Calling the API without credentials fails."""
+    _map = {  # METHOD: ( {_REQUEST_PARAMS} , {_EXPECTED_RESULTS}
+        "GET": ({"method": "GET", "url": path, "params": params}, {200}),
+        "HEAD": ({"method": "HEAD", "url": path}, {200}),
+    }
+    if method in _map.keys():
+        _request, _results = _map.get(method)
+        resp = client.request(**_request)
+        assert resp.status_code in _results
```
Comment (jwhitlock) [inline tests/unit/test_monitor.py:38]: No, this copies the API tests without making any of the changes needed to test the monitoring endpoints. The test is called ``_fails`` but tests that they return 200s, tests with an authenticated client even though API credentials are not needed, and for some reason passes ``{"params": None}`` for a few cases.

Let's do this:

```suggestion
@pytest.mark.parametrize("method", ("HEAD", "GET"))
@pytest.mark.parametrize("path", ("/__lbheartbeat", "/__heartbeat"))
def test_monitoring_endpoints(anon_client, method, path):
    """Monitoring endpoints can be called without credentials"""
    resp = anon_client.request(method, path)
    assert resp.status_code == 200
```

Comment (jwhitlock) [reply tests/unit/test_monitor.py:38]: but, the existing tests do pass and test the new methods.

I'm touching the monitoring tests for issue #153, I can make my changes then if you want to merge and move on to the next issue.

File: tests/unit/test_monitor.py [hunk 5]
```diff
@@ -2,37 +2,9 @@
 
 import pytest
```
Comment (bsieber-mozilla) [inline tests/unit/test_monitor.py:3]: _Going to hop to branch to clean this_

Comment (bsieber-mozilla) [reply tests/unit/test_monitor.py:1]: from `db_session` to `dbsession`

**Comment by bsieber-mozilla** [comment]
@jwhitlock this should resolve, then I'll merge it in

**Pull request merged**